src/dev.rs

```diff
@@ -18,9 +22,27 @@ mod graph;
 #[cfg_attr(docsrs, doc(cfg(feature = "dev-graph")))]
 pub use graph::{circuit_dot_graph, layout::circuit_layout};
 
+/// Cells that haven't been explicitly assigned to, default to zero.
+fn cell_value<F: Field>(cell: Option<F>) -> F {
+    cell.unwrap_or_else(F::zero)
+}
+
 /// The reasons why a particular circuit is not satisfied.
 #[derive(Debug, PartialEq)]
 pub enum VerifyFailure {
+    /// A cell used in an active gate was not assigned to.
+    Cell {
+        /// The column in which this cell is located.
+        column: Column<Any>,
+        /// The row in which this cell is located.
+        row: usize,
+        /// The index of the active gate. These indices are assigned in the order in which
+        /// `ConstraintSystem::create_gate` is called during `Circuit::configure`.
+        gate_index: usize,
+        /// The name of the active gate. These are specified by the gate creator (such as
+        /// a chip implementation), and may not be unique.
+        gate_name: &'static str,
+    },
```

Comment:
Could we have this be a Warning instead of a Failure? If the assignment passes all the constraints, it technically is still _a_ satisfying assignment (just maybe not the intended one)

Reply:
I'm inclined to keep this as a failure for now; if someone needs this case to occur, they can explicitly handle that in their tests. We don't currently have a way to distinguish between warnings and errors without moving away from the `Result` type, so I'd rather think this through in a separate issue.

Reply:
I think it's completely fine for this to be an error. If the value doesn't matter then the circuit should explicitly assign 0, IMHO. Detecting potential circuit errors is more important than the minor convenience of being able to leave some cells (that are used in a gate) unassigned.